collect_app/src/main/java/org/odk/collect/android/activities/FormFillingActivity.java

```diff
@@ -2109,6 +2110,7 @@ private void finishAndReturnInstance() {
     }
 
     private void exit() {
+        ReferenceManager.instance().reset();
```

Comment:
Another thought: do we actually need to use JavaRosa's XForm parsing to get the data we need out of the XForm? If parsing works the first time, it'd imply to me that we don't care about actually having a setup `ReferenceManager` when downloading a form. If that's the case, we could just write our own XML parse code (or expose a subset of JavaRosa's) to grab the form metadata.

Reply:
> Could we instead reset the ReferenceManager before parsing the form when downloading?

Yes but:
1. We do the reset when opening a form so it's more natural to me to reset it on exiting as well. We don't need to keep those references after exiting.
2. It would be impossible to write a test that is failing first (the same I've added) as we need to parse a form before adding it to our `StubOpenRosaServer` so the test would fail there. I would need to reset the manager before doing that first parsing but after doing that it would be enough and during downloading the failure would not take place.

> This also makes me wonder if there are weird interactions between form download and form entry given they share a ReferenceManager instance. Ideally we want to remove the static state here (it's a blocker for multiple form entry sessions as well), and I'm also wondering if we should take the hit to solve this issue.

Yes currently it's weird but we wouldn't need to parse the whole form with external datsets to read the metadata.

> Another thought: do we actually need to use JavaRosa's XForm parsing to get the data we need out of the XForm? If parsing works the first time, it'd imply to me that we don't care about actually having a setup ReferenceManager when downloading a form. If that's the case, we could just write our own XML parse code (or expose a subset of JavaRosa's) to grab the form metadata.

As I said above we shouldn't need to parse the whole form and it would be better to have a parser that only reads the metadata we need. But I think it could be a separate issue focused on improving the process of downloading forms rather than fixing the issue as resetting the manager is enough.

Reply:
All good points! I'll have a think about what we might wnat to plan for the next release, but it definitely feels like something about `ReferenceManager` and or the parsing needs to change.